### PR TITLE
Validators list page validator card

### LINKED ISSUE #4327: Validators list - Validators cards
## Context

- https://github.com/Joystream/pioneer/issues/4321

## Design

[Link to Figma> ](https://www.figma.com/file/tjad3r2EZJXRMFOELcxxNR/Validators?node-id=5%3A1550&t=jqHGNwf0jVupH08w-1)

Design Preview

## Scope
- [ ] Add validator cards for known members (validator stash account is bound to membership)
- [ ] Add validator cards for unknown members
- [ ] Add pagination

### PR DESCRIPTION
**Pull request by eshark9312**
- #4327 

~~This PR is blocked by #4404~~
Add Validator Card component

[> Design](https://www.figma.com/file/QnjZuWYtGimTEkQlN6MNlH/Validators?type=design&node-id=1799-410079&mode=design&t=CTRyskvdyi8y9VYk-0)
[> Story](https://pioneer-2-storybook-git-fork-eshark9312-valida-23fd7f-joystream.vercel.app/?path=/story/pages-validators-validatorlist--statistics-and-lists)

### CODE REVIEW

**Review by chrlschwb — APPROVED** [review]

**Comment by thesan** [comment]
@chrlschwb this PR is not ready for review it includes a lot of changes from #4404 because git isn't able to differentiate identify them in the "squash and merge" commit (so they show as conflicts). @eshark9312 IMO you should `git rebase --onto dev validators-list-filter-search validators-list-page-validatorCard` but after you make sure that your local `dev` is at the latest version from `Joystream/pioneer`.

File: packages/ui/src/validators/modals/validatorCard/ValidatorDetail.tsx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+              }
+            />
+          </RowGapBlock>
+          <SidePaneRow>
+            <SidePaneLabel text="Email" />
+            <SidePaneText>Alice1@joystream.com</SidePaneText>
+          </SidePaneRow>
+          <SidePaneRow>
+            <SidePaneLabel text="Website" />
+            <SidePaneText>Alice1joystream.com</SidePaneText>
```
Comment (thesan) [inline packages/ui/src/validators/modals/validatorCard/ValidatorDetail.tsx:60]: I think these sections (About, Email, Website, and all the other external resources) come from the membership bound to the validator account. 

Thus there is no need to re-implement these fields instead please re-use the code from Membership pane.

As a result I imagine this should be implemented after #4428 is merged.

File: packages/ui/src/app/pages/Validators/ValidatorList.stories.tsx [hunk 2]
```diff
@@ -113,9 +120,13 @@ export default {
                   { who: 'j4RxTMa1QVucodYPfQGA2JrHxZP944dfJ8qdDDYKU4QbJCWNP', value: joy(0.2) },
                   { who: 'j4Rxkb1w9yB6WXroB2npKjRJJxwxbD8JjSQwMZFB31cf5aZAJ', value: joy(0.2) },
                   { who: 'j4RyLBbSUBvipuQLkjLyUGeFWEzmrnfYdpteDa2gYNoM13qEg', value: joy(0.2) },
+                  { who: 'j4S998Thq5kQHyurofh8QfHrcFN2c1T19gTdMGUVVx5EHKgky', value: joy(0.2) },
                 ],
               },
               erasTotalStake: joy(130_000),
+              slashingSpans: {
+                size: Math.floor(Math.random() * 2),
```
Comment (thesan) [inline packages/ui/src/app/pages/Validators/ValidatorList.stories.tsx:128]: For this mock to work `withUnwrap` could be redefined like this:
```ts
const withUnwrap = (data: Record<any, any>) =>
  Object.defineProperties(data, {
    unwrap: { value: () => data },
    isSome: { value: Object.keys(data).length > 0 },
  })
```
This way defining an empty object would simulate a Nothing.

File: packages/ui/src/validators/modals/validatorCard/ValidatorDetail.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+              <Stat size="s" value={validator.isVerified ? 'Verified' : 'Unverified'}>
+                <TextSmall lighter>Status</TextSmall>
+              </Stat>
+              <Stat size="s" value={validator.slashed.toString()}>
+                <TextSmall lighter>Slashed</TextSmall>
+              </Stat>
+              <Stat size="s" value={0}>
+                <TextSmall lighter>Uptime</TextSmall>
+              </Stat>
+              <Stat size="s" value={validator.APR.toString()}>
```
Comment (thesan) [inline packages/ui/src/validators/modals/validatorCard/ValidatorDetail.tsx:38]: This should have a `%` after it

File: packages/ui/src/validators/modals/validatorCard/ValidatorDetail.tsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+              <Stat size="s" value={20000}>
+                <TextSmall lighter>Total reward</TextSmall>
+              </Stat>
+              <Stat size="s" value={validator.isVerified ? 'Verified' : 'Unverified'}>
+                <TextSmall lighter>Status</TextSmall>
+              </Stat>
+              <Stat size="s" value={validator.slashed.toString()}>
+                <TextSmall lighter>Slashed</TextSmall>
+              </Stat>
+              <Stat size="s" value={0}>
```
Comment (thesan) [inline packages/ui/src/validators/modals/validatorCard/ValidatorDetail.tsx:35]: This too. Also do you know how to get this value ?

File: packages/ui/src/validators/modals/validatorCard/ValidatorDetail.tsx [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+export const ValidatorDetail = ({ validator }: Props) => {
+  return (
+    <>
+      <SidePaneBody>
+        <Details gap={24}>
+          <RowGapBlock gap={4}>
+            <h6>Key elements</h6>
+            <ModalStatistics>
+              <Stat size="s" value={20000}>
```
Comment (thesan) [inline packages/ui/src/validators/modals/validatorCard/ValidatorDetail.tsx:26]: This should show as an amount of JOY (although its' not in the design) so it should use the `TokenValue` component somewhere (maybe you could use `TokenValueStat`.
Also do you know how to get this value ?

File: packages/ui/src/validators/modals/validatorCard/ValidatorDetail.tsx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                <TextSmall lighter>Slashed</TextSmall>
+              </Stat>
+              <Stat size="s" value={0}>
+                <TextSmall lighter>Uptime</TextSmall>
+              </Stat>
+              <Stat size="s" value={validator.APR.toString()}>
+                <TextSmall lighter>Average APR</TextSmall>
+              </Stat>
+              <Stat size="s" value={validator.staking.others.length}>
+                <TextSmall lighter>Nominators</TextSmall>
```
Comment (thesan) [inline packages/ui/src/validators/modals/validatorCard/ValidatorDetail.tsx:42]: In the design this is  "Staked by nominators":
image

And this should also show as an amount of JOY

File: packages/ui/src/validators/modals/validatorCard/ValidatorDetail.tsx [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+          <RowGapBlock gap={4}>
+            <h6>Key elements</h6>
+            <ModalStatistics>
+              <Stat size="s" value={20000}>
+                <TextSmall lighter>Total reward</TextSmall>
+              </Stat>
+              <Stat size="s" value={validator.isVerified ? 'Verified' : 'Unverified'}>
+                <TextSmall lighter>Status</TextSmall>
+              </Stat>
+              <Stat size="s" value={validator.slashed.toString()}>
```
Comment (thesan) [inline packages/ui/src/validators/modals/validatorCard/ValidatorDetail.tsx:32]: This should have "times" or "time" at the end.

You could use:
```suggestion
              <Stat size="s" value={`${validator.slashed} time${plural(validator.slashed)}`}>
```

**Review by thesan — CHANGES_REQUESTED** [review]

File: packages/ui/src/validators/modals/validatorCard/ValidatorDetail.tsx [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+export const ValidatorDetail = ({ validator }: Props) => {
+  return (
+    <>
+      <SidePaneBody>
+        <Details gap={24}>
+          <RowGapBlock gap={4}>
+            <h6>Key elements</h6>
+            <ModalStatistics>
```
Comment (thesan) [inline packages/ui/src/validators/modals/validatorCard/ValidatorDetail.tsx:25]: Also please match the [design](https://www.figma.com/file/QnjZuWYtGimTEkQlN6MNlH/Validators?type=design&node-id=1799-410079&mode=design&t=CTRyskvdyi8y9VYk-0) order for the statistics.

image

**Comment by thesan** [comment]
@eshark9312 `eshark9312:update-validators-list-columns` was merged into this branch [here](https://github.com/Joystream/pioneer/pull/4419/commits/7065554e9d651beaad3d7c601f31b5b9ed15fa27) has a result this PR now depends on #4643.

#4643 is almost ready, so as soon as it gets merged into `dev` please merge `dev` into this PR. It will make the review a lot easier.

File: packages/ui/src/mocks/helpers/asChainData.ts [hunk 9]
```diff
@@ -20,4 +23,9 @@ export const asChainData = (data: any): any => {
   }
 }
 
-const withUnwrap = (data: Record<any, any>) => Object.defineProperty(data, 'unwrap', { value: () => data })
+const withUnwrap = (data: Record<any, any>) =>
+  Object.defineProperties(data, {
+    unwrap: { value: () => data },
+    isSome: { value: Object.keys(data).length > 0 },
+    get: { value: (key: AccountId32) => data[encodeAddress(key.toString())] },
```
Comment (thesan) [inline packages/ui/src/mocks/helpers/asChainData.ts:30]: Good idea but just to make this a bit more general:
```suggestion
    get: {
      value: (key: unknow) => {
        switch (key) {
          case 'object':
            if (key.toRawType?.() === 'AccountId') {
              return data[encodeAddress(key.toString())]
            }
          default:
            return data[key.toString()]
        }
      }
    }
```

**Review by thesan — CHANGES_REQUESTED** [review]

**Review by thesan — APPROVED** [review]
Great work! The graph looks good, I think we should go with `chart.js` instead of `nivo` for the financial dashboard too.

FYI all the nominate buttons will be changed into links to the the Polkadot app to nominate validators. But this should be done in a follow up PR. Also I don't remember what the copy should be.

**Pull request merged**